csharp_dotnetcore/11.QnAMaker/Startup.cs

```diff
@@ -0,0 +1,164 @@
+
+            // Initialize Bot Connected Services Clients
+            var connectedServices = InitBotServices(botConfig);
+            services.AddSingleton(sp => connectedServices);
+
+            services.AddSingleton(sp => botConfig);
+
+            services.AddBot<QnABot>(options =>
+            {
+                options.CredentialProvider = new ConfigurationCredentialProvider(Configuration);
+
+                // Add MyAppInsightsLoggerMiddleware (logs activity messages into Application Insights)
+                var appInsightsLogger = new MyAppInsightsLoggerMiddleware(connectedServices.TelemetryClient.InstrumentationKey, logUserName: true, logOriginalMessage: true);
+                options.Middleware.Add(appInsightsLogger);
+            });
+        }
+
+        /// <summary>
+        /// This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
+        /// </summary>
+        /// <param name="app">The application builder.  This provides the mechanisms to configure an application's request pipeline.</param>
+        /// <param name="env">Provides information about the web hosting environment an application is running in.</param>
+        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
+        {
+            if (env.IsDevelopment())
+            {
+                app.UseDeveloperExceptionPage();
+            }
+
+            app.UseDefaultFiles()
+                .UseStaticFiles()
+                .UseBotFramework();
+        }
+
+        /// <summary>
+        /// Initialize the bot's references to external services.
+        ///
+        /// For example, Application Insights and QnaMaker services
+        /// are created here.  These external services are configured
+        /// using the BotConfigure class (based on the contents of your ".bot" file).
+        /// </summary>
+        /// <param name="config">Configuration object based on your ".bot" file.</param>
+        /// <returns>A object representing client objects to access external services the bot uses.</returns>
+        private static BotServices InitBotServices(BotConfiguration config)
+        {
+            var connectedServices = new BotServices();
+
+            foreach (var service in config.Services)
+            {
+                switch (service.Type)
+                {
+                    case ServiceTypes.QnA:
+                        {
+                            // Create a QNA Maker that is initialized and suitable for passing
+                            // into the IBot-derived class (QnABot).
+                            // In this case, we're creating a custom class (wrapping the original
+                            // QnAMaker client) that logs the results of QnA Maker into Application
+                            // Insights for future anaysis.
+                            var qna = service as QnAMakerService;
```

Comment:
If something was misconfigured, this "as" statment will result in the qna variable being null, and the next line will cryptically throw a null reference exception. Consider making this a hard cast, so the exception that results is clear and from the right line. 
var qna = (QnaMakerService)service;

Reply:
... or checking for null, and explicitly throwing with a clear exception. That's probably the "cleanest" thing to do.